### ISSUE TITLE
lxc.loglevel and lxc.logfile seem to be ignored

### ISSUE THREAD

**Issue by brauner** [lxc/lxc#429]
Specifying

```
lxc.loglevel = 1
lxc.logfile = /home/chb/.local/share/lxc/vivid/container.log
```

in an unprivileged containers `config` does not seem to have any effect. Meaning I get no file called `container.log` but only a file named after the container (in this case `vivid.log`) which only logs ERROR and not DEBUG. The container has been started with:

```
lxc-start -n vivid -F


Kernel 3.18.5
Archlinux
lxc 1.1
cgmanager 0.35
lxcfs 0.5
```

**Comment by cburroughs** [comment]
I'm seeing this behavior with containers started as root with lxc 1.0.7.  The `--logpriority` cmd line flag does work.

**Comment by matthijskooijman** [comment]
Same here with 1.0.6 on Debian Jessie.

Looking at the code for 1.0.6 (and using a debugger), I found that `lxc-start`'s main function calls [`lxc_log_options_no_override()`](https://github.com/lxc/lxc/blob/lxc-1.0.6/src/lxc/log.c#L443), which locks the log options and prevents further changes, *before* parsing the config file. When the config file is then parsed, the logfile parsed is silently discared [here]( https://github.com/lxc/lxc/blob/lxc-1.0.6/src/lxc/log.c#L423).

In git master, the [function is still called](https://github.com/lxc/lxc/blob/master/src/lxc/lxc_start.c#L232), but it no longer attemps to lock the log file, but only the log level now: https://github.com/lxc/lxc/blob/master/src/lxc/log.c#L461-L465

However, the `lxc_loglevel_specified` variable there seems to be only used in `lxc_log_init()` in a way that doesn't really make sense to me. Perhaps one of the developers can comment about how they think this should work?

**Comment by matthijskooijman** [comment]
Looking more closely at the history, I believe this was broken in 6edbfc86508fb07d2b8377dce40f36f3032c32a5

Before that commit, the level and filename could only be set once. If a filename was passed to `lxc_log_init` (e.g. from the commandline), that would count as that one time, and any corresponding directive in the config file would be ignored. But when nothing was specified on the commandline, I believe the entries in the config file would be that one time, so they would take effect normally.

That commit attempts to allow `lxc_log_set_file()` and `lxc_log_set_level()` to be called more than once, as part of the API and introduces the `lxc_log_options_no_override()` in an attempt to preserve the original behaviour for the lxc commandline binaries. However, that latter didn't work out, the behaviour now changed to always discarding the settings from the config file, and only using the commandline options (AFAICS, at least).

In git master, some changes have been made to the logfile handling for multithreading, but I believe that made the `lxc_log_options_override()` stuff even more inconsistent than it already was.

**Comment by hallyn** [comment]
Note that this works fine in newer lxc.  This was a bit hairy to get right,
so I'd believe there are some ranges of trees where it's broken.

Can you reproduce this with any of the stable-xx trees?

**Comment by matthijskooijman** [comment]
Hm, it seemed to me that the code in git master wouldn't actually work either, though I mostly didn't really understood how it works. Looking more closely now, it seems that the override stuff in git master only takes effect if `lxc_log_options_no_override()` is specified *after* `lxc_log_init()` is called, which I suppose never happens? Also, setting the log options through e.g. `lxc_log_set_level` seems to always work now. I wonder if that means that options on the commandline no longer override options in the config file?

Anyway, I'll see if I can  test a newer version. Might take a while though.

Thanks!

**Comment by evgeni** [comment]
I think this is fixed since 2.0.0, @brauner @hallyn let's close?

**Comment by matthijskooijman** [comment]
Seems this is indeed working on 2.0.9 for me. Thanks! :-)
